src/frontend/js/hooks/useUnionResource/utils/fetchEntity.ts

```diff
@@ -0,0 +1,73 @@
+import { QueryClient } from '@tanstack/query-core';
+import { PaginatedResourceQuery, PaginatedResponse } from 'types/Joanie';
+import { Maybe } from 'types/utils';
+import { REACT_QUERY_SETTINGS } from 'settings';
+import { EOF_QUERY_KEY } from './constants';
+
+export type FetchDataFunction<Data, FetchDataFilter> = (
+  filters: FetchDataFilter,
+) => Promise<PaginatedResponse<Data>>;
+
+export interface QueryConfig<Data, Filters> {
+  queryKey: string[];
+  fn: FetchDataFunction<Data, Filters>;
+  filters: Filters;
+}
+
+export interface FetchEntityBaseArgs {
+  page: number;
+  perPage: number;
+  eofRef: React.MutableRefObject<Record<string, number>>;
+  queryClient: QueryClient;
+}
+interface FetchEntityArgs<Data, Filters> extends FetchEntityBaseArgs {
+  queryConfig: QueryConfig<Data, Filters>;
+}
+export const fetchEntity = async <
+  Data,
+  Filters extends PaginatedResourceQuery = PaginatedResourceQuery,
+>({
+  queryConfig,
+  page,
+  perPage,
+  eofRef,
+  queryClient,
+}: FetchEntityArgs<Data, Filters>): Promise<Maybe<PaginatedResponse<Data>>> => {
+  const { queryKey, fn, filters: queryConfigFilters } = queryConfig;
+  const queryKeyString = queryKey.join('-');
+  // Do not fetch if we already reached the end of the list.
+  if (typeof eofRef.current[queryKeyString] === 'number' && page > eofRef.current[queryKeyString]) {
+    return;
+  }
+
+  const filters: Filters = {
+    ...(queryConfigFilters || {}),
+    page,
+    page_size: perPage,
+  };
+  const QUERY_KEY = ['user', ...queryKey, JSON.stringify(filters)];
```

Comment:
As this method is generic, it should not add `user` in for all queries.
```suggestion
  const QUERY_KEY = [...queryKey, JSON.stringify(filters)];
```